### ISSUE TITLE
Is it possible to access a DVC S3 Remote with intake?

### ISSUE THREAD

**Issue by cwerner** [intake/intake#595]
Hi all. 👋

I'm currently using intake to access some datafiles (xarray, csv, shp) from a remote storage (our on-premise S3 service). This works fine, but I now wonder if it's possible to also use an intake catalog to access to current state from a dvc remote store on S3?
DVC offers a Python API to work with files so I figured this should/ could also work via an intake plugin? 

```
import dvc.api

with dvc.api.open(
        'get-started/data.xml',
        repo='https://github.com/iterative/dataset-registry'
        ) as fd:
    # ... fd is a file descriptor that can be processed normally.
``` 

Has anyone worked with this setup? I haven't written my own plugins yet so any help/ pointer would be welcome!

Cheers

**Comment by martindurant** [comment]
I'm not entirely sure what you mean, but DVC is now using fsspec internally ( https://github.com/iterative/dvc/issues/5622 ), and fsspec also has an experimental [dvc backend](https://github.com/intake/filesystem_spec/blob/master/fsspec/implementations/dvc.py) - so you could format a file path as "dvc://...". 

From an Intake point of view, you are thinking of having a list of datasets based on DVC?

**Comment by cwerner** [comment]
Thanks for the response @martindurant 

Sorry if I was not clear... Yes, I want to define a couple of datasets in an intake catalog. These datasets currently live in a plain s3 bucket and I can access them no problem.
Now, I instead want to access them in a dvc remote s3 bucket (since they are generated by a pipeline and will change in the future). Here data is stored with hashed names and one has to use dvc to access the files based on their "true" filename and not the cryptic hash-key names.

Hope this makes sense?

Anyway, seems like this might be handled by fsspec if they can handle a dvc:// URL. I'll try to read at fsspec!

**Comment by martindurant** [comment]
Yes, changing your URL to dvc:// is the way to do it, but the driver at fsspec is very rudimentary, so you will probably have to develop it some more. Since DVC is now plugged in with fsspec, their devs may be interested in helping you figure this out (cc @isidentical )

**Comment by cwerner** [comment]
Thanks. From the GitHub issues I found I get the impression that this functionality is currently worked on/ will change quite a bit in the near future? Can this be used right now or should I wait and watch these issues for future updates? 

Thanks again

**Comment by martindurant** [comment]
For fsspec's dvc: driver,
"Use at your own risk"! I don't know of someone is actively working on it. Please report on any attempt you make.

**Comment by efiop** [comment]
Hi @cwerner !

Thanks for the interest! As @martindurant mentioned, we are migrating to fsspec right now (thanks to @isidentical 's great work and help from @martindurant :pray: ). We already have our old DvcFileSystem implementation https://github.com/iterative/dvc/blob/master/dvc/fs/dvc.py that we use ourselves inside dvc and that will be migrated to fully comply with fsspec soon (we expect it to finalise migration in ~4 weeks) and we'll then try to register it as the official implementation, that other people could use. So please stay tuned! :satellite:

**Comment by cwerner** [comment]
Thanks for the heads-up @efiop 

I'll watch this space. Big fan of DVC, too 👍

**Comment by martindurant** [comment]
@isidentical : are you or a colleague interested in this topic?

**Comment by isidentical** [comment]
We are still migrating our remotes one by one, so first we need to finish them. After that, as @efiop said we'll work on the DVC fsspec implementation.

**Comment by observingClouds** [comment]
Hi,

I just came across this issue today and am glad to see the progress that has been made since the initial question. I couldn't find any good documentation about the fsspec dvc protocol, but https://github.com/iterative/dvc/issues/9943 has been a good starting point. The good news is that it is now indeed possible to combine dvc, s3 and intake.

The following catalog syntax works for me:
```yaml
plugins:
  source:
    - module: intake_xarray
sources:
  dataset1:
    driver: netcdf
    description: Dataset description
    args:
      chunks: null
      urlpath: dvc://relative/path/to/dataset1.nc  # path in git repository
      storage_options:
        url: https://github.com/user/git-dvc-repository
        rev: abcde1234567  # optional git commit hash to pin specific version
```

What is really neat about this in my opinion is that:
- versions can be pinned to specific version or always point to the most recent version (`rev`)
- the versions are defined by git-commits rather than md5sums of individual files (although this has the downside that several git-versions can point to the same dataset if this particular dataset has not changed)
- if one would like to use md5sums instead of the git-commits this is also possible if the dataset is a single file by
  ```yaml
    dataset1:
    driver: netcdf
    description: Dataset pinned by md5sum
    args:
      chunks: null
      urlpath: s3://<bucket>/files/md5/13/5223bb12e5773bb23bd19c2b231298
      storage_options:
        anon: false
  ```
  and circumventing the dvc pipeline.
  The md5sum can be retrieved from e.g. `dvc.lock` or files like `dataset1.nc.dvc`.
- The s3-bucket information (or potentially any other dvc supported remote) is provided in the `.dvc/config` file and handled by dvc and not mentioned in the intake catalog.
  ```
   [core]
      remote = aws
   ['remote "aws"']
      url = s3://<bucket>
  ```

**What currently does not work** based on my limited knowledge here is the pinning of datasets that consist of several files, like the zarr-format. In principle it shouldn't be too hard to implement this feature as well because dvc saves all necessary information in `md5/hash/ofdirectory.dir` files that could serve as a basis to create a reference filesystem. These files look like
```json
{"md5": "d2d464bf8dbf870b9afd18c279c6fd51", "relpath": ".zattrs"}, 
{"md5": "e20297935e73dd0154104d4ea53040ab", "relpath": ".zgroup"}, 
{"md5": "6e427adf518ac592295453cf814c9087", "relpath": ".zmetadata"}, 
{"md5": "2d9208a22001378fd91a7e8e903a7708", "relpath": "temp/.zarray"}, 
{"md5": "b61893298ae5782db071b180559d6a57", "relpath": "temp/.zattrs"},
 {"md5": "3d16fcb9a54ac64d9770581cd55c9a03", "relpath": "temp/0.0.0"}, 
{"md5": "a956910ad8d032a2b9b261f12462b244", "relpath": "temp/0.0.1"}
```
@martindurant would the fsspec `reference` protocol allow a file like this as is (with a template file) or would one need to write a new driver?

Curious to hear any suggestions.

**Comment by martindurant** [comment]
@efiop you might find this use case interesting

**Comment by martindurant** [comment]
Answering this specifically

> @martindurant would the fsspec reference protocol allow a file like this as is (with a template file) or would one need to write a new driver?

Yes! You could store this "manifest" in the git repo itself or elsewhere, and you would effectively have a third type of FS to access your data through, alongside dvc: and s3:. Since a referenceFS can contain references to multiple target FSs, you can even combine them, so that you can address by gitref or hash or whatever, per entry. You *could* have each any chunk appear multiple times.

The structure of a reference set is JSON is easy enough to make, at its simplest:
```
{"name": ["target-path"]}
```
e.g.,
```
{
".zmetadata": ["s3://<bucket>/files/md5/13/5223bb12e5773bb23bd19c2b231298"],
...
}
```
so you can make this structure manually.

Because the referenceFS needs a number of parameters on creation (essentially, the arguments to each target filesystem), *that* is something you definitely want to hide in an intake catalog.
